storage/src/main/java/io/delta/storage/internal/S3LogStoreUtil.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (2022) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.storage.internal;
+
+import com.amazonaws.services.s3.model.ListObjectsV2Request;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.RemoteIterator;
+import org.apache.hadoop.fs.s3a.Listing;
+import org.apache.hadoop.fs.s3a.S3AFileStatus;
+import org.apache.hadoop.fs.s3a.S3AFileSystem;
+import org.apache.hadoop.fs.s3a.S3ListRequest;
+
+import java.io.IOException;
+import java.util.HashSet;
+
+import static org.apache.hadoop.fs.s3a.S3AUtils.ACCEPT_ALL;
+import static org.apache.hadoop.fs.s3a.S3AUtils.iteratorToStatuses;
+
+
+/**
+ * Static utility methods for the S3SingleDriverLogStore.
+ *
+ * Used to trick the class loader so we can use methods of org.apache.hadoop:hadoop-aws without needing to load this as
+ * a dependency for tests in core.
+ */
+public class S3LogStoreUtil {
+    private S3LogStoreUtil() {}
+
+    /**
+     * Uses the S3ListRequest.v2 interface with the startAfter parameter to only list files
+     * which are lexicographically greater than resolvedPath.
+     */
+    public static FileStatus[] s3ListFrom(FileSystem fs, Path resolvedPath, Path parentPath) throws IOException {
+        S3AFileSystem s3afs = (S3AFileSystem) fs;
```

Comment:
I have not tried this, but don't fully understand how this would work. We need to know the type of the filesystem, since we use methods which are specific to S3AFileSystem. Unfortunately, even if some class have an identical implementation of S3AFileSystem, we would have no way to access its methods. For instance, the getListing method used below is particular to S3AFileSystem, and not part of any interface or super class of hadoop-common.

Reply:
ping @zsxwing

Reply:
You can write code like this:
```
if (fs.getClass.getName == "org.apache.hadoop.fs.s3a.S3AFileSystem") {
   S3AFileSystem s3afs = (S3AFileSystem) fs;
   your new impl
} else {
   the previous old impl we have today
}
```

> even if some class have an identical implementation of S3AFileSystem

I don't get this point. I guess I was not clear. There are multiple S3 file system implementations today. `S3AFileSystem` is the most popular one but people may use others. `(S3AFileSystem) fs` will fail when the user doesn't use `S3AFileSystem`.

Reply:
Ah, thanks for clarifying. Should we use this instead of the feature flag, or in combination with the feature flag?

Reply:
ping @zsxwing

Reply:
we should still have a feature flag.

Reply:
Great. Currently we have a feature flag, but not a check on the class name. Will it suffice to only have a feature flag, or should we have the class name check in addition?

Reply:
both :) if you turn on the feature flag, but you aren't using the right S3AFileSystem, things won't work well. let's prevent users from shooting themselves in the foot.

thanks for these great questions and for staying on top of this!

Reply:
Added in https://github.com/delta-io/delta/pull/1210/commits/8843feec90a04d70bcb3c72a0c4e05243e7c46fc

This prevents the user from shooting themselves in the foot, but can leave the user with a gun they think is working, while in fact it has no bullets.

Reply:
then perhaps we should throw an error? to let them know that something is wrong.

Reply:
great analogies :P

Reply:
> then perhaps we should throw an error? to let them know that something is wrong.

If we only have the feature flag and not the class name check we will get a class cast exception if the feature flag is enabled but the S3AFileSystem implementation is wrong (i.e., pre https://github.com/delta-io/delta/commit/8843feec90a04d70bcb3c72a0c4e05243e7c46fc ). We could throw a different exception instead?

Reply:
Done in https://github.com/delta-io/delta/pull/1210/commits/fc0caeadd4a463a3c0e6fe68ae3042ee7a47bc8d